packages/context/src/context.ts

```diff
@@ -34,16 +52,142 @@ export class Context {
+    super();
     if (typeof _parent === 'string') {
       name = _parent;
       _parent = undefined;
     }
     this._parent = _parent;
     this.name = name || uuidv1();
+
+    this.setupEventHandlers();
+  }
+
+  /**
+   * Wrap the debug statement so that it always print out the context name
+   * as the prefix
+   * @param formatter
+   * @param args
+   */
+  // tslint:disable-next-line:no-any
+  private _debug(formatter: any, ...args: any[]) {
+    /* istanbul ignore if */
+    if (debug.enabled) {
+      if (typeof formatter === 'string') {
+        debug(`[%s] ${formatter}`, this.name, ...args);
+      } else {
+        debug('[%s] ', this.name, formatter, ...args);
+      }
+    }
+  }
+
+  /**
+   * Set up an internal listener to notify registered observers asynchronously
+   * upon `bind` and `unbind` events
+   */
+  private setupEventHandlers() {
+    // The following are two async functions. Returned promises are ignored as
+    // they are long-running background tasks.
+    this.startNotificationTask('bind').catch(err => {
+      // Catch error to avoid lint violations
+      debug(err);
+      this.emit('error', err);
+    });
+    this.startNotificationTask('unbind').catch(err => {
+      // Catch error to avoid lint violations
+      debug(err);
+      this.emit('error', err);
+    });
+  }
+
+  /**
+   * Start a background task to listen on context events and notify observers
+   * @param eventType Context event type
+   */
+  private async startNotificationTask(eventType: ContextEventType) {
+    const notificationEvent = `${eventType}-notification`;
+    this.on(eventType, binding => {
+      // No need to schedule notifications if no observers are present
+      if (this.observers.size === 0) return;
+      // Track pending events
+      this.pendingEvents++;
+      // Take a snapshot of current observers to ensure notifications of this
+      // event will only be sent to current ones. Emit a new event to notify
+      // current context observers.
+      this.emit(notificationEvent, {
+        binding,
+        observers: new Set(this.observers),
+      });
+    });
+    // FIXME(rfeng): p-event should allow multiple event types in an iterator.
+    // Create an async iterator from the given event type
+    const events: AsyncIterable<{
+      binding: Readonly<Binding<unknown>>;
+      observers: Set<ContextObserver>;
+    }> = pEvent.iterator(this, notificationEvent);
+    for await (const event of events) {
+      // The loop will happen asynchronously upon events
+      try {
+        // The execution of observers happen in the Promise micro-task queue
+        const binding = event.binding;
+        await this.notifyObservers(eventType, binding, event.observers);
+        this.pendingEvents--;
+        this._debug(
+          'Observers notified for %s of binding %s',
+          eventType,
+          binding.key,
+        );
+        this.emit('observersNotified', {eventType, binding});
```

Comment:
This allocates a new object for each event triggered. It would be more performant to use multiple positional arguments instead:

```ts
this.emit('observersNotified', eventType, binding);
```

Having said that, I don't know how well this works with `p-event` and whether the performance impact is significant enough.

Reply:
`p-event` allows `multiArgs` but it ends up creating `[eventType, binding]` as the return value from the promise. I think the overhead is similar.